src/trace/ord/key_batch.rs

```diff
@@ -0,0 +1,432 @@
+        }
+
+        // if self.lower1 == self.upper1 || self.lower2 == self.upper2 {
+        //     // these are just copies, so let's bite the bullet and just do them.
+        //     if self.lower1 < self.upper1 { self.result.copy_range(&source1.layer,
+        // self.lower1, self.upper1); self.lower1 = self.upper1; }     if self.
+        // lower2 < self.upper2 { self.result.copy_range(&source2.layer, self.lower2,
+        // self.upper2); self.lower2 = self.upper2; } }
+        // Merging is complete; only copying remains. Copying is probably faster than
+        // merging, so could take some liberties here.
+        if self.lower1 == self.upper1 || self.lower2 == self.upper2 {
+            // Limit merging by remaining fuel.
+            let remaining_fuel = *fuel - effort;
+            if remaining_fuel > 0 {
+                if self.lower1 < self.upper1 {
+                    let mut to_copy = remaining_fuel as usize;
+                    if to_copy < 1_000 {
+                        to_copy = 1_000;
+                    }
+                    if to_copy > (self.upper1 - self.lower1) {
+                        to_copy = self.upper1 - self.lower1;
+                    }
+                    self.result
+                        .copy_range(&source1.layer, self.lower1, self.lower1 + to_copy);
+                    self.lower1 += to_copy;
+                }
+                if self.lower2 < self.upper2 {
+                    let mut to_copy = remaining_fuel as usize;
+                    if to_copy < 1_000 {
+                        to_copy = 1_000;
+                    }
+                    if to_copy > (self.upper2 - self.lower2) {
+                        to_copy = self.upper2 - self.lower2;
+                    }
+                    self.result
+                        .copy_range(&source2.layer, self.lower2, self.lower2 + to_copy);
+                    self.lower2 += to_copy;
+                }
+            }
+        }
+
+        effort = (self.result.vals.vals.len() - starting_updates) as isize;
+
+        *fuel -= effort;
+
+        // if *fuel < -1_000_000 {
+        //     eprintln!("Massive deficit OrdKey::work: {}", fuel);
+        // }
+    }
+}
+
+/// A cursor for navigating a single layer.
+#[derive(Debug)]
+pub struct OrdKeyCursor<T: Lattice + Ord + Clone, R: MonoidValue, O = usize> {
+    valid: bool,
+    empty: (),
+    cursor: OrderedCursor<OrderedLeaf<T, R>>,
+    phantom: PhantomData<O>,
+}
+
+impl<K, T, R, O> Cursor<K, (), T, R> for OrdKeyCursor<T, R, O>
+where
+    K: Ord + Clone,
+    T: Lattice + Ord + Clone,
+    R: MonoidValue,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    type Storage = OrdKeyBatch<K, T, R, O>;
+
+    fn key<'a>(&self, storage: &'a Self::Storage) -> &'a K {
+        self.cursor.key(&storage.layer)
+    }
+    fn val<'a>(&self, _storage: &'a Self::Storage) -> &'a () {
+        unsafe { ::std::mem::transmute(&self.empty) }
```

Comment:
This is UB, replace it with a reference to a unit literal (it's a literal so it gets static promotion)
```suggestion
        &()
```